### PR TITLE
Purity and Completeness of PFPs

### PR DESCRIPTION
**Pull request by henrylay97**
This PR adds the functionality to fill the completeness and purity created in SBNSoftware/sbnanaobj#29.

For now it continues to use the Track based functions for tracks and showers although a `GetShowerPrimary` method is employed to roll up showers if they have not been simulated using roll up. If a particle has been rolled up (or it is not a shower-type particle) this function will return the original ID. A map based method is employed to ensure this doesn't add significant time penalties to CAFMaker.

I have validated that the plots of the completeness and purity are consistent with those created from existing LArSoft analyzer modules used in the CI.

### CODE REVIEW

File: sbncode/CAFMaker/FillTrue.cxx [hunk 1]
```diff
@@ -972,27 +994,35 @@ float ContainedLength(const TVector3 &v0, const TVector3 &v1,
...
 
   ret.total_deposited_energy = total_energy / 1000. /* MeV -> GeV */;
 
   // setup the matches
   for (auto const &pair: matches) {
-    caf::ParticleMatch match;
+    caf::SRParticleMatch match;
     match.G4ID = pair.first;
     match.energy = pair.second / 1000. /* MeV -> GeV */;
+    match.hit_purity = (hits.size() != 0) ? (track_hits_map.find(match.G4ID)->second).first / (float) hits.size() : 0.;
```
Comment (etyley) [inline sbncode/CAFMaker/FillTrue.cxx:1016]: Whilst very concise, `->second.first` is not the most readable code. The current implementation also requires running `find` multiple times for each map which could carry some overhead when there is a large number of particles (i.e. cosmic events with no roll up).  
I would suggest running find once and saving the output in an appropriately named variable.

File: sbncode/CAFMaker/RecoUtils/RecoUtils.cc [hunk 2]
```diff
@@ -9,6 +9,7 @@ std::vector<std::pair<int, float>> CAFRecoUtils::AllTrueParticleIDEnergyMatches(
     for (unsigned int idIt = 0; idIt < trackIDs.size(); ++idIt) {
       int id = trackIDs[idIt].trackID;
       if (rollup_unsaved_ids) id = std::abs(id);
+      id = GetShowerPrimary(id);
```
Comment (etyley) [inline sbncode/CAFMaker/RecoUtils/RecoUtils.cc:12]: Should this be encapsulated by the `rollup_unsaved_ids` if statement?

File: sbncode/CAFMaker/CMakeLists.txt [hunk 3]
```diff
@@ -31,6 +31,7 @@ art_make_library( LIBRARY_NAME sbncode_CAFMaker
                   larcore_Geometry_Geometry_service
                   larcorealg_GeoAlgo
                   larsim_MCCheater_BackTrackerService_service
+                  larsim_Utils
```
Comment (etyley) [inline sbncode/CAFMaker/CMakeLists.txt:34]: I can't see where any includes for this were added?

**Review by etyley — APPROVED** [review]
Looks good to me, some suggestions on how to improve readability but nothing major

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 4]
```diff
@@ -776,6 +776,7 @@ void CAFMaker::produce(art::Event& evt) noexcept {
   // Prep truth-to-reco-matching info
   std::map<int, std::vector<std::pair<geo::WireID, const sim::IDE*>>> id_to_ide_map = PrepSimChannels(simchannels, *geometry);
   std::map<int, std::vector<art::Ptr<recob::Hit>>> id_to_truehit_map = PrepTrueHits(hits, clock_data, *bt_serv.get());
+  std::map<int, std::pair<int, float>> id_to_hit_energy_map = SetupIDHitEnergyMap(hits, clock_data, *bt_serv.get());
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:779]: I think I am echoing Ed by suggesting that this `std::pair` becomes something like `struct HitsEnergy{int nhits; float totE;};`

**Review by cjbackhouse — APPROVED** [review]

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 4 above)
Comment (henrylay97) [reply sbncode/CAFMaker/CAFMaker_module.cc]: I like this idea, have followed it

File: sbncode/CAFMaker/FillTrue.cxx (hunk 1 above)
Comment (henrylay97) [reply sbncode/CAFMaker/FillTrue.cxx]: Thanks Ed, definitely a good suggestion :)

File: sbncode/CAFMaker/FillTrue.cxx [hunk 5]
```diff
@@ -972,27 +994,40 @@ float ContainedLength(const TVector3 &v0, const TVector3 &v1,
...
     match.G4ID = pair.first;
     match.energy = pair.second / 1000. /* MeV -> GeV */;
+
+    caf::HitsEnergy track_matched_hits = track_hits_map.find(match.G4ID)->second;
+    caf::HitsEnergy all_matched_hits = all_hits_map.find(match.G4ID)->second;
+
+    match.hit_purity = (hits.size() != 0) ? track_matched_hits.nHits / (float) hits.size() : 0.;
+    match.energy_purity = (ret.total_deposited_energy > std::numeric_limits<float>::epsilon()) ? match.energy / ret.total_deposited_energy : 0.;
+    match.hit_completeness = (all_matched_hits.nHits != 0) ? track_matched_hits.nHits / (float) all_matched_hits.nHits : 0.;
+    match.energy_completeness = (all_matched_hits.totE > std::numeric_limits<float>::epsilon()) ? pair.second / all_matched_hits.totE : 0.;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FillTrue.cxx:1023]: Are these tests against epsilon just due to paranoia about floating point comparisons? Test against zero is safe.
```suggestion
    match.hit_purity = (hits.size() != 0) ? track_matched_hits.nHits / (float) hits.size() : 0.;
    match.energy_purity = (ret.total_deposited_energy > 0) ? match.energy / ret.total_deposited_energy : 0.;
    match.hit_completeness = (all_matched_hits.nHits != 0) ? track_matched_hits.nHits / (float) all_matched_hits.nHits : 0.;
    match.energy_completeness = (all_matched_hits.totE > 0) ? pair.second / all_matched_hits.totE : 0.;
```

File: sbncode/CAFMaker/CMakeLists.txt (hunk 3 above)
Comment (etyley) [reply sbncode/CAFMaker/CMakeLists.txt:34]: Not sure how I missed those sorry :/

File: sbncode/CAFMaker/FillTrue.cxx (hunk 5 above)
Comment (henrylay97) [reply sbncode/CAFMaker/FillTrue.cxx:1023]: Yeah they are a little paranoid when you put it like that!

**Pull request merged**